### ISSUE TITLE
ODP | ODB: /createUploadSession If-Match should not generate a 412 Precondition Failed when the exact online eTag is being used

### ISSUE THREAD

**Issue by abraunegg** [OneDrive/onedrive-api-docs#1716]
#### Category
- [ ] Question
- [X] Documentation issue
- [ ] Bug

#### Expected or Desired Behavior

When using a session to upload a modified file, the API should not present a `412 Precondition Failed` error when using `If-Match` and the provided eTag equals the existing online eTag.

https://learn.microsoft.com/en-us/onedrive/developer/rest-api/api/driveitem_createuploadsession

#### Observed Behavior

1. Query the exact file item online to get the absolute latest eTag item, compare against current known value
2. Attempt to create an upload session via `/createUploadSession` and add the `If-Match` header with the known eTag value
3. Get a 412 response from OneDrive API

```
[DEBUG] Setting file permissions for: /home/alex/.config/onedrive/refresh_token
[DEBUG] Request URL = https://graph.microsoft.com/v1.0/drives/66d53be8a5056eca/root:/4mb_word_doc%2FH2A%20Research%20Equivalence%20v6.docx
*   Trying 20.190.142.172...
* Connected to graph.microsoft.com (20.190.142.172) port 443 (#1)
* ALPN, offering http/1.1
* successfully set certificate verify locations:
*   CAfile: /etc/pki/tls/certs/ca-bundle.crt
  CApath: none
* SSL connection using TLSv1.3 / TLS_AES_256_GCM_SHA384
* ALPN, server did not agree to a protocol
* Server certificate:
*  subject: C=US; ST=Washington; L=Redmond; O=Microsoft Corporation; CN=graph.microsoft.com
*  start date: Jun 28 00:00:00 2023 GMT
*  expire date: Jun 28 23:59:59 2024 GMT
*  subjectAltName: host "graph.microsoft.com" matched cert's "graph.microsoft.com"
*  issuer: C=US; O=DigiCert Inc; CN=DigiCert SHA2 Secure Server CA
*  SSL certificate verify ok.
> GET /v1.0/drives/66d53be8a5056eca/root:/4mb_word_doc%2FH2A%20Research%20Equivalence%20v6.docx HTTP/1.1
Host: graph.microsoft.com
User-Agent: ISV|abraunegg|OneDrive Client for Linux/v2.5.0-dev
Accept: */*
Authorization: bearer <Redacted>

< HTTP/1.1 200 OK
< Cache-Control: no-store
< Transfer-Encoding: chunked
< Content-Type: application/json;odata.metadata=minimal;odata.streaming=true;IEEE754Compatible=false;charset=utf-8
< Strict-Transport-Security: max-age=31536000
< request-id: a0d14097-1ff0-4889-a028-ba2c5a98e780
< client-request-id: a0d14097-1ff0-4889-a028-ba2c5a98e780
< x-ms-ags-diagnostic: {"ServerInfo":{"DataCenter":"Australia Southeast","Slice":"E","Ring":"4","ScaleUnit":"000","RoleInstance":"ML1PEPF00004F6B"}}
< OData-Version: 4.0
< Date: Mon, 31 Jul 2023 06:26:14 GMT
< 
[DEBUG] onedrive.performHTTPOperation() => OneDrive HTTP Server Response: 200
* Connection #1 to host graph.microsoft.com left intact
[DEBUG] curlEngine.http.perform() => HTTP Response Headers: ["strict-transport-security":"max-age=31536000", "x-ms-ags-diagnostic":"{\"ServerInfo\":{\"DataCenter\":\"Australia Southeast\",\"Slice\":\"E\",\"Ring\":\"4\",\"ScaleUnit\":\"000\",\"RoleInstance\":\"ML1PEPF00004F6B\"}}", "transfer-encoding":"chunked", "odata-version":"4.0", "date":"Mon, 31 Jul 2023 06:26:14 GMT", "client-request-id":"a0d14097-1ff0-4889-a028-ba2c5a98e780", "request-id":"a0d14097-1ff0-4889-a028-ba2c5a98e780", "cache-control":"no-store", "content-type":"application/json;odata.metadata=minimal;odata.streaming=true;IEEE754Compatible=false;charset=utf-8"]
[DEBUG] OneDrive API Response: {"@microsoft.graph.downloadUrl":"https:\/\/mwg7ww.ph.files.1drv.com\/y4mo9rtLgLlSUh88_XuyOnu9UVYuvrcfHwi_1e1dviDdiM_NMmWNsSvmjTZ0Hb6wp4tB0bjS18DFsL5naN9IkHcc30D-v9RUiQw6jR1thO8Ry4frPbDm7sDklS8P6oOy7mG7PulwuLMbiOkUfFtUvnQLuHoLI2MChDgk4e71BTVWaQuyusyTdSRsSyH9dHp9NZPfiY0ws9CTnRBDwR56Mm1tnz0Y2z9ytY1PVjOHyjZVIE","@odata.context":"https:\/\/graph.microsoft.com\/v1.0\/$metadata#drives('66d53be8a5056eca')\/root\/$entity","cTag":"aYzo2NkQ1M0JFOEE1MDU2RUNBITI4NTA1Mi4xOTM1","createdBy":{"application":{"displayName":"OneDrive Client for Linux","id":"4035b959"},"user":{"displayName":"SADF ASDF","id":"66d53be8a5056eca"}},"createdDateTime":"2023-07-31T05:40:17.047Z","eTag":"aNjZENTNCRThBNTA1NkVDQSEyODUwNTIuOQ","file":{"hashes":{"quickXorHash":"Ml4zGCvJpAyqzLM5FaZWhpGLqs0=","sha1Hash":"310BE58EFA885921A012FD84D35313FB371F35FF","sha256Hash":"E7D28D034AE2233450267A9DD39AFD9EDD1FB785459575C55D7FEE4E6DE39FE9"},"mimeType":"application\/vnd.openxmlformats-officedocument.wordprocessingml.document"},"fileSystemInfo":{"createdDateTime":"2023-07-31T05:40:17.046Z","lastModifiedDateTime":"2023-07-31T05:41:33.956Z"},"id":"66D53BE8A5056ECA!285052","lastModifiedBy":{"user":{"displayName":"SADF ASDF","id":"66d53be8a5056eca"}},"lastModifiedDateTime":"2023-07-31T05:41:37.05Z","name":"H2A Research Equivalence v6.docx","parentReference":{"driveId":"66d53be8a5056eca","driveType":"personal","id":"66D53BE8A5056ECA!285051","name":"4mb_word_doc","path":"\/drive\/root:\/4mb_word_doc"},"reactions":{"commentCount":0},"size":4316411,"webUrl":"https:\/\/1drv.ms\/w\/s!AMpuBaXoO9VmkbJ8"}
[DEBUG] dbItem.eTag:            aNjZENTNCRThBNTA1NkVDQSEyODUwNTIuOQ
[DEBUG] currentOnlineData eTag: aNjZENTNCRThBNTA1NkVDQSEyODUwNTIuOQ
* Found bundle for host graph.microsoft.com: 0x1c0c270 [can pipeline]
* Re-using existing connection! (#1) with host graph.microsoft.com
* Connected to graph.microsoft.com (20.190.142.172) port 443 (#1)
> POST /v1.0/drives/66d53be8a5056eca/items/66D53BE8A5056ECA!285052:/H2A%20Research%20Equivalence%20v6.docx:/createUploadSession HTTP/1.1
Host: graph.microsoft.com
User-Agent: ISV|abraunegg|OneDrive Client for Linux/v2.5.0-dev
Accept: */*
If-Match: aNjZENTNCRThBNTA1NkVDQSEyODUwNTIuOQ
Content-Type: application/json
Authorization: bearer <Redacted>
Content-Length: 110

* We are completely uploaded and fine
< HTTP/1.1 412 Precondition Failed
< Cache-Control: no-store
< Transfer-Encoding: chunked
< Content-Type: application/json
< Strict-Transport-Security: max-age=31536000
< request-id: 1db71ddc-ae37-4563-8e1f-a136d065e35c
< client-request-id: 1db71ddc-ae37-4563-8e1f-a136d065e35c
< x-ms-ags-diagnostic: {"ServerInfo":{"DataCenter":"Australia Southeast","Slice":"E","Ring":"4","ScaleUnit":"000","RoleInstance":"ML1PEPF00004F6B"}}
< Date: Mon, 31 Jul 2023 06:26:14 GMT
< 
[DEBUG] onedrive.performHTTPOperation() => OneDrive HTTP Server Response: 412
* Connection #1 to host graph.microsoft.com left intact
[DEBUG] curlEngine.http.perform() => HTTP Response Headers: ["strict-transport-security":"max-age=31536000", "x-ms-ags-diagnostic":"{\"ServerInfo\":{\"DataCenter\":\"Australia Southeast\",\"Slice\":\"E\",\"Ring\":\"4\",\"ScaleUnit\":\"000\",\"RoleInstance\":\"ML1PEPF00004F6B\"}}", "transfer-encoding":"chunked", "date":"Mon, 31 Jul 2023 06:26:14 GMT", "client-request-id":"1db71ddc-ae37-4563-8e1f-a136d065e35c", "request-id":"1db71ddc-ae37-4563-8e1f-a136d065e35c", "cache-control":"no-store", "content-type":"application/json"]
[DEBUG] Exit scope called
onedrive.OneDriveException@src/onedrive.d(1154): HTTP request returned status code 412 (Precondition Failed)
{
    "error": {
        "code": "resourceModified",
        "innerError": {
            "client-request-id": "1db71ddc-ae37-4563-8e1f-a136d065e35c",
            "date": "2023-07-31T06:26:15",
            "request-id": "1db71ddc-ae37-4563-8e1f-a136d065e35c"
        },
        "message": "ETag does not match current item's value"
    }
}
```

#### Steps to Reproduce

As per above

Thank you.

[ ]: http://aka.ms/onedrive-api-issues
[x]: http://aka.ms/onedrive-api-issues

**Comment by abraunegg** [comment]
@ificator 
Could you look at this issue with some priority?

This is a long standing API issue, where the only work around is to delete the file online so that the locally modified file can be uploaded to OneDrive.

The problem here is that any versioning is now also lost - because the file online was deleted.

**Comment by ificator** [comment]
I'll see if I can get a local repro - in the meantime, what are you providing in the body of the `createUploadSession` request?

**Comment by abraunegg** [comment]
@ificator 
The request body is conforming with this: https://learn.microsoft.com/en-us/onedrive/developer/rest-api/api/driveitem_createuploadsession?view=odsp-graph-online#request-body

![image](https://github.com/OneDrive/onedrive-api-docs/assets/4956234/a158b43d-3fc1-4073-83a1-e19b057f54e5)


**Example:**
```
{
	"item": {
		"@microsoft.graph.conflictBehavior": "replace",
		"fileSystemInfo": {
			"lastModifiedDateTime": "2023-07-31T19:49:08Z"
		}
	}
}
```

**Comment by ificator** [comment]
Thanks - and just to tick the obvious boxes, if you make another get call after the failed POST to createUploadSession is the etag still the one you provided?

**Comment by abraunegg** [comment]
@ificator 
> Thanks - and just to tick the obvious boxes, if you make another get call after the failed POST to createUploadSession is the etag still the one you provided?

I have double checked - the eTag is the same online post the session create failure.
```
[DEBUG] Request URL = https://graph.microsoft.com/v1.0/drives/66d53be8a5056eca/root:/4mb_word_doc%2FH2A%20Research%20Equivalence%20v6-onedrive-client-dev.docx
[DEBUG] dbItem.eTag:            aNjZENTNCRThBNTA1NkVDQSEyODUwNTQuNg
[DEBUG] currentOnlineData eTag: aNjZENTNCRThBNTA1NkVDQSEyODUwNTQuNg
POST BODY: {"item":{"@microsoft.graph.conflictBehavior":"replace","fileSystemInfo":{"lastModifiedDateTime":"2023-07-31T20:51:47Z"}}}
OneDrive returned a 'HTTP 412 - Precondition Failed' - gracefully handling error
uploadSessionData: {"error":{"code":"resourceModified","innerError":{"client-request-id":"bb9e17a2-1eee-4da1-979e-f8d582b3bafe","date":"2023-07-31T20:54:05","request-id":"bb9e17a2-1eee-4da1-979e-f8d582b3bafe"},"message":"ETag does not match current item's value"}}
[DEBUG] Request URL = https://graph.microsoft.com/v1.0/drives/66d53be8a5056eca/root:/4mb_word_doc%2FH2A%20Research%20Equivalence%20v6-onedrive-client-dev.docx
[DEBUG] dbItem.eTag:            aNjZENTNCRThBNTA1NkVDQSEyODUwNTQuNg
[DEBUG] currentOnlineData post sessionCreate attempt eTag: aNjZENTNCRThBNTA1NkVDQSEyODUwNTQuNg
```

**Comment by ificator** [comment]
Actually I see something that's a little odd in your request and I can repro when I mimic it:

```
POST /v1.0/drives/66d53be8a5056eca/items/66D53BE8A5056ECA!285052:/H2A%20Research%20Equivalence%20v6.docx:/createUploadSession HTTP/1.1
```

In your request (above for easy reference) the item wiht id `285052` is the one you want to update, but then the path is also provided. It looks like the store returns a 412 when the target item is not found but a precondition was specified, which is why you're getting a 412 instead of a 404. Without the precondition I'm guessing you'd get some kind of "you cannot create a file under the provided parent".

TLDR: remove the file name from your update request.

**Comment by abraunegg** [comment]
@ificator 
The `/createUploadSession` request however is conforming to the docs:

![image](https://github.com/OneDrive/onedrive-api-docs/assets/4956234/6edf837b-f49e-4d98-8e1b-257f83bfd2d0)

The data that is being used is as follows:
```
string url = driveByIdUrl ~ parentDriveId ~ "/items/" ~ parentId ~ ":/" ~ encodeComponent(filename) ~ ":/createUploadSession";
```
so with this:
```
/v1.0/drives/66d53be8a5056eca/items/66D53BE8A5056ECA!285052:/H2A%20Research%20Equivalence%20v6.docx:/createUploadSession
```

parentDriveId = 66d53be8a5056eca
parentId = 66D53BE8A5056ECA!285052 (which is a directory called '4mb_word_doc')
filename = encoded filename

If I remove the 'If-Match' from creating the session to upload a locally modified file, the `/createUploadSession` request generates a 403:

```
[DEBUG] Using Curl defaults for HTTP operational protocol version (potentially HTTP/2)
[DEBUG] Setting file permissions for: /home/alex/.config/onedrive/refresh_token
[DEBUG] Request URL = https://graph.microsoft.com/v1.0/drives/66d53be8a5056eca/root:/4mb_word_doc%2FH2A%20Research%20Equivalence%20v6-onedrive-client-dev.docx
[DEBUG] dbItem.eTag:            aNjZENTNCRThBNTA1NkVDQSEyODUwNTQuNg
[DEBUG] currentOnlineData eTag: aNjZENTNCRThBNTA1NkVDQSEyODUwNTQuNg
POST BODY: {"item":{"@microsoft.graph.conflictBehavior":"replace","fileSystemInfo":{"lastModifiedDateTime":"2023-07-31T20:51:47Z"}}}
OneDrive returned a 'HTTP 403 - Forbidden' - gracefully handling error
[DEBUG] Exit scope called
onedrive.OneDriveException@src/onedrive.d(1153): HTTP request returned status code 403 (Forbidden)
{
    "error": {
        "code": "accessDenied",
        "innerError": {
            "client-request-id": "9e2477ab-dee0-47f3-a17b-27ca39e6fe17",
            "date": "2023-07-31T21:15:48",
            "request-id": "9e2477ab-dee0-47f3-a17b-27ca39e6fe17"
        },
        "message": "Access Denied"
    }
}
```

**Comment by abraunegg** [comment]
@ificator 
And, stripping the filename from the createSession request actually fails to then upload a new or modified file:
```
createUploadSession URL: https://graph.microsoft.com/v1.0/drives/66d53be8a5056eca/items/66D53BE8A5056ECA!285055/createUploadSession
POST BODY: {"item":{"@microsoft.graph.conflictBehavior":"replace","fileSystemInfo":{"lastModifiedDateTime":"2011-01-19T09:56:00Z"}}}
Uploading new file ./4mb_word_doc/H2A Research Equivalence v6.docx ... failed.

ERROR: Microsoft OneDrive API returned an error with the following message:
  Error Message:    HTTP request returned status code 400 (Bad Request)
  Error Reason:     File name not provided in url.
  Error Timestamp:  2023-07-31T21:25:22
  API Request ID:   2db299d2-ef7e-4dfa-a1af-29bcf4833fec
  Calling Function: performNewFileUpload()

Uploading new file ./4mb_word_doc/H2A Research Equivalence v6.docx ... failed.
```

**Comment by abraunegg** [comment]
@ificator 
**New File Upload via session:**
```
createUploadSession URL: https://graph.microsoft.com/v1.0/drives/66d53be8a5056eca/items/66D53BE8A5056ECA!285055:/H2A%20Research%20Equivalence%20v6.docx:/createUploadSession
POST BODY: {"item":{"@microsoft.graph.conflictBehavior":"replace","fileSystemInfo":{"lastModifiedDateTime":"2011-01-19T09:56:00Z"}}}
Uploading 100% |oooooooooooooooooooooooooooooooooooooooo| DONE IN 00:00:04                                                                                                           
Uploading new file ./4mb_word_doc/H2A Research Equivalence v6.docx ... done.
```

**Modified file (post successful upload): **
```
The file content has changed locally and has a newer timestamp, thus needs to be uploaded to OneDrive
Changed local items to upload to OneDrive: 1

createUploadSession URL: https://graph.microsoft.com/v1.0/drives/66d53be8a5056eca/items/66D53BE8A5056ECA!285056:/H2A%20Research%20Equivalence%20v6.docx:/createUploadSession
POST BODY: {"item":{"@microsoft.graph.conflictBehavior":"replace","fileSystemInfo":{"lastModifiedDateTime":"2023-07-31T21:29:57Z"}}}
onedrive.OneDriveException@src/onedrive.d(1163): HTTP request returned status code 412 (Precondition Failed)
{
    "error": {
        "code": "resourceModified",
        "innerError": {
            "client-request-id": "6b0a2651-0618-4987-8988-bcefcbf48995",
            "date": "2023-07-31T21:30:18",
            "request-id": "6b0a2651-0618-4987-8988-bcefcbf48995"
        },
        "message": "ETag does not match current item's value"
    }
}


```

**Comment by abraunegg** [comment]
@ificator 
I think this is potentially coming down to a documentation interpretation bug:

![image](https://github.com/OneDrive/onedrive-api-docs/assets/4956234/91504539-b23a-481b-a6ac-e4a54b0d53e3)

What is occurring - is that when a file is modified, the session attempting to be created uses that items id - as per the document extract above

This creates a request URL of the following, that generates the 412 error:
```
https://graph.microsoft.com/v1.0/drives/66d53be8a5056eca/items/66D53BE8A5056ECA!285056:/H2A%20Research%20Equivalence%20v6.docx:/createUploadSession
```

66D53BE8A5056ECA!285056 is the ID of the file to update.

Now - when I change this to use the parent ID, the request URL is:
```
https://graph.microsoft.com/v1.0/drives/66d53be8a5056eca/items/66D53BE8A5056ECA!285055:/H2A%20Research%20Equivalence%20v6.docx:/createUploadSession
```

And this generates the required JSON for session upload ...... and no 412 Precondition Failed error ....

**Comment by abraunegg** [comment]
@ificator 
Re-Flagging this as a documentation issue

**Comment by ificator** [comment]
Here's the PR - please take a look and let me know if you thing there's still improvements that could be made:
https://github.com/microsoftgraph/microsoft-graph-docs/pull/22299

**Comment by abraunegg** [comment]
@ificator 

> Here's the PR - please take a look and let me know if you thing there's still improvements that could be made: [microsoftgraph/microsoft-graph-docs#22299](https://github.com/microsoftgraph/microsoft-graph-docs/pull/22299)

That certainly reads much better and the examples are much clearer as well

**Comment by abraunegg** [comment]
Closing this as done
